### PR TITLE
[llvm-objdump] Fix coloring with nested WithMarkup

### LINKED ISSUE #99661: [BUG] Weird coloring in x86 assembly in llvm-objdump
Hi,
I am experimenting a little with writing my own backend (for another project).
To test the correctness of code generatation and object file emiting i use `llvm-mc` and `llvm-objdump`.
I dissasembled an file and i noticed a bracket is green and the closing one is gray.
I think it would be prettier to use gray for both ones:
image

(You maybe need to zoom in a little to view the difference)

### PR DESCRIPTION
**Pull request by MaskRay**
WithMarkup objects may nest, resulting in the `)` in `leaq
(%rdx,%rax), %rbx` to be green instead of the default color,
mismatching the color of `(`.

```
% llvm-mc -triple=x86_64 -mdis <<< '0x48 0x8d 0x1c 0x02'
        .text
        leaq    <mem:(<reg:%rdx>,<reg:%rax>)>, <reg:%rbx>
```

To ensure that `(` and `)` get the same color, maintain a color stack
within MCInstPrinter.

Fix #99661

### CODE REVIEW

File: llvm/lib/MC/MCInstPrinter.cpp [hunk 1]
```diff
@@ -226,27 +226,32 @@ format_object<uint64_t> MCInstPrinter::formatHex(uint64_t Value) const {
 
 MCInstPrinter::WithMarkup MCInstPrinter::markup(raw_ostream &OS,
                                                 Markup S) const {
-  return WithMarkup(OS, S, getUseMarkup(), getUseColor());
+  return WithMarkup(const_cast<MCInstPrinter &>(*this), OS, S, getUseMarkup(),
```
Comment (jh7370) [inline llvm/lib/MC/MCInstPrinter.cpp:229]: Overall approach looks fine, but this does feel a bit like an abuse of `const`. Could `const` be dropped from the `markup` method?

Comment (MaskRay) [reply llvm/lib/MC/MCInstPrinter.cpp:229]: Many const methods transitively call the `const` `markup`. If we want to drop `const_cast` here, many target functions `printXXX` will need be adjusted...

Comment (jh7370) [reply llvm/lib/MC/MCInstPrinter.cpp:229]: Yuck, okay. I think that's probably still the "right" thing to do (after all you're modifying the state of the class when calling these methods in this manner). But I acknowledge that's probably a bit impractical too, so if you add a TODO comment here instead, I'd be happy with that.

I also considered making the `ColorStack` member `mutable`, but that doesn't really feel any better than the `const_cast` in this context.

**Review by jh7370 — APPROVED** [review]

**Comment by Cr0a3** [comment]
Thank you for fixing this

File: llvm/lib/MC/MCInstPrinter.cpp (hunk 1 above)
Comment (MaskRay) [reply llvm/lib/MC/MCInstPrinter.cpp]: Migrated `printRegName` users. Dropped `const_cast` now.

**Pull request merged**